### PR TITLE
feat(cdc) EventsV2 dataset

### PR DESCRIPTION
**Pull request by fpacifici**
This is a joined dataset between groupedmessage and events. It is based on a LEFT JOIN from groups to events (we can optimize this join and revisit the exact structure when we try to fit it into the product).
This PR introduces:
- The dataset itself that depends on the schemas of the existing datasets and joins them
- A subclass of ColumnSet to be used for hierarchical column sets
- A concept of table_alias to be passed to datasets when running column processing, so that dataset specific column processing (column_expr) can produce results that can be successfully used in a joined query without making events know about the join.
- Fixes the way escaping is done since it would not work with aliases that includes a "."
- A big unit test for all column processing logic.

Test plan:
- the existing coverage for column_expr and query processing still works
- ran the tests from sentry.
- added test_joined_columns that reproduces most of the existing column processing tests, but on a joined dataset with table aliases. See that test for all corner cases
- ran a series of queries. Example:

```
{
    "selected_columns": ["events.tags[level]", "events.event_id", "events.issue", "events.tags_key", "groups.first_release_id"],
    "conditions": [["groups.id", "=", 1395]],
    "project": [2],
    "from_date": "2019-09-21T18:55:09",
    "to_date": "2019-09-27T18:55:09",
    "granularity": 3600
}

RESPONSE
{
    "data": [
        {
            "events.tags[level]": "error",
            "events.event_id": "ad39dbd35ebe4ca7a182b883143dd460",
            "events.issue": 1395,
            "events.tags_key": "level",
            "groups.first_release_id": 8
        },
        {
            "events.tags[level]": "error",
            "events.event_id": "ad39dbd35ebe4ca7a182b883143dd460",
            "events.issue": 1395,
            "events.tags_key": "runtime",
            "groups.first_release_id": 8
        }
    ],
    "meta": [
        {
            "name": "events.tags[level]",
            "type": "Nullable(String)"
        },
        {
            "name": "events.event_id",
            "type": "FixedString(32)"
        },
        {
            "name": "events.issue",
            "type": "Nullable(UInt64)"
        },
        {
            "name": "events.tags_key",
            "type": "String"
        },
        {
            "name": "groups.first_release_id",
            "type": "Nullable(UInt64)"
        }
    ],
....


```

### CODE REVIEW

File: snuba/datasets/eventsv2.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    SchemaJoinedSource,
+    SubJoinSource,
+)
+from snuba.query.extensions import PERFORMANCE_EXTENSION_SCHEMA, PROJECT_EXTENSION_SCHEMA
+from snuba.query.schema import GENERIC_QUERY_SCHEMA
+from snuba.request import RequestSchema
+from snuba.schemas import get_time_series_extension_properties
+
+
+class EventsV2(TimeSeriesDataset):
```
Comment (fpacifici) [inline snuba/datasets/eventsv2.py]: This is kind of experimental for now, not sure if the composite dataset we will use in production will be called eventsV2

File: snuba/datasets/eventsv2.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                JoinCondition(
+                    left=JoinConditionExpression(table_alias="groups", column="project_id"),
+                    right=JoinConditionExpression(table_alias="events", column="project_id"),
+                ),
+                JoinCondition(
+                    left=JoinConditionExpression(table_alias="groups", column="id"),
+                    right=JoinConditionExpression(table_alias="events", column="group_id"),
+                ),
+            ],
+            join_type=JoinType.LEFT,
```
Comment (fpacifici) [inline snuba/datasets/eventsv2.py]: Will revisit the type of join when we will start building the event search on top of this.

File: snuba/datasets/eventsv2.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        dataset_schemas = DatasetSchemas(
+            read_schema=schema,
+            write_schema=None,
+        )
+        super().__init__(
+            dataset_schemas=dataset_schemas,
+            processor=None,
+            default_topic=None,
+            time_group_columns={
+                'events.time': 'events.timestamp',
```
Comment (fpacifici) [inline snuba/datasets/eventsv2.py]: I am hesitating whether we should refer to columns within a join as string with an alias inside (which is consistent with the way we address columns today. Except for when we build a schema, they are just strings). Or to build an abstraction on them, but that has big implications on everything that access the columns of a dataset.

File: snuba/datasets/factory.py [hunk 4]
```diff
@@ -25,11 +25,13 @@ def get_dataset(name):
     from snuba.datasets.cdc.groupedmessage import GroupedMessageDataset
     from snuba.datasets.transactions import TransactionsDataset
     from snuba.datasets.outcomes import OutcomesDataset
+    from snuba.datasets.eventsv2 import EventsV2
     dataset_mappings = {
         'events': EventsDataset,
         'groupedmessage': GroupedMessageDataset,
         'transactions': TransactionsDataset,
         'outcomes': OutcomesDataset,
+        'eventsv2': EventsV2,
```
Comment (fpacifici) [inline snuba/datasets/factory.py]: I am surprised this worked at the first attempt without circular depndencies

File: snuba/datasets/eventsv2.py (hunk 1 above)
Comment (markstory) [reply snuba/datasets/eventsv2.py]: Should we be concerned about reusing terms between sentry and snuba? `eventsv2` in sentry is ~= to discover2.

Comment (fpacifici) [reply snuba/datasets/eventsv2.py]: discoverv1.5 ?

File: snuba/datasets/table_storage.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+from datetime import datetime
+
+from snuba.clickhouse import DATETIME_FORMAT
+from snuba.datasets.schemas.tables import WritableTableSchema
+from snuba.processor import MessageProcessor
+from snuba.writer import BatchWriter
+
+
+class TableWriter:
```
Comment (fpacifici) [inline snuba/datasets/table_storage.py]: I introduced this to avoid the type issues whenI had to set a None processor and a None default topic.
This is a step towards splitting dataset (which represent the datamodel) and the TableStorage which represent the table. 
At this point this contains only what is needed to write on the table but going on it should evolve this way:
1) split Clickhouse schema from snuba schema
2) move Clickhouse schema here
3) rename it to table storage
4) move some single table dataset logic into here
5) keep the snuba schema and the query logic (that runs the join) into the dataset.

File: snuba/datasets/eventsv2.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        schema = JoinedSchema(join_structure)
+        dataset_schemas = DatasetSchemas(
+            read_schema=schema,
+            write_schema=None,
+        )
+        super().__init__(
+            dataset_schemas=dataset_schemas,
+            processor=None,
+            default_topic=None,
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/datasets/eventsv2.py]: See the new TableWriter. I doubt we could go further within this PR

**Comment by fpacifici** [comment]
Will split the TableWriter out of the dataset

File: snuba/datasets/eventsv2.py (hunk 3 above)
Comment (fpacifici) [reply snuba/datasets/eventsv2.py]: I changed it to keep the structure in the ColumnSet. 
Anyway this list (besides DDL which is irrelevant here) is used only to find if a column passed in the query (thus unstructured) is in the columnSet.
We will be able to do better when we will change the way we parse our query and once we will give more structure to the query object.

File: snuba/datasets/__init__.py [hunk 7]
```diff
@@ -25,8 +25,10 @@ class Dataset(object):
     def __init__(self,
             dataset_schemas: DatasetSchemas,
             *,
-            processor: MessageProcessor,
-            default_topic: str,
+            processor: Optional[MessageProcessor],
+            default_topic: Optional[str],
+            # PR 479 changes the architecture of the dataset
```
Comment (fpacifici) [inline snuba/datasets/__init__.py]: All this change will automatically disappear when PR 479 gets committed and we will have  fewer nullable fields out of datasets.

File: snuba/datasets/groups.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        else:
+            return (match[1], match[2])
+
+    def column_expr(self, column_name, body):
+        aliased_column = self.__parse_qualified_column(column_name)
+        table_alias = aliased_column[0]
+        if table_alias == self.GROUPS_ALIAS:
+            return self.__grouped_message.column_expr(column_name, body)
+        else:
+            # this allows to delegate columns like tags[blabla] to
```
Comment (fpacifici) [inline snuba/datasets/groups.py]: I think the right way to solve this is to change the way we parse the query and split the logic that concerns formatting from the logic that concerns parsing and transformation:
Example:
- right after json validation we build the Query object (as we do today)
- the query object will contain an abstract list of selected columns expressed as an AST of dataclasses parsed from the string (instead of the bare strings we have today).
- dataset can provide some custom dataclasses to use (Tags for example)
- The, query organized this way, goes through the query processors that can edit it.
- The query is passed to the Clickhouse query class and the formatting logic happens there (example transforming a FunctionCall class into function(x). This logic, today is conflated into column_expr with more abstract query editing.

This all anyway, is not needed for this dataset that can delegate column_expr to events.

File: snuba/clickhouse/columns.py [hunk 9]
```diff
@@ -302,3 +304,23 @@ def get(self, key, default=None):
 
     def for_schema(self):
         return ', '.join(column.for_schema() for column in self.columns)
+
+
+class QualifiedColumnSet(ColumnSet):
```
Comment (fpacifici) [inline snuba/clickhouse/columns.py]: This preserves some structure in the ColumnSet when we have columns coming from different tables.
It still allows iteration the way we did before (which is what is needed in util.conditions_expr) but it does not loose the which column comes from which table as the previous implementation did.

File: snuba/datasets/groups.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+from snuba.query.extensions import (
+    PerformanceExtension,
+    ProjectExtension,
+    QueryExtension,
+)
+from snuba.query.query import AliasedColumn
+from snuba.query.timeseries import TimeSeriesExtension
+
+
+class Groups(TimeSeriesDataset):
```
Comment (fpacifici) [inline snuba/datasets/groups.py]: Naming keeps being really really hard.
The idea behind this name is that it joins groups and assignees with events. So generally we start with groups.
Overall it will evolve to the whole datamodel, but at this point I don;t want to call it in a more generic way.

**Comment by tkaemming** [comment]
Holding off on re-reviewing this until #479 is in.

**Comment by fpacifici** [comment]
#479 merged and brought here

File: snuba/datasets/groups.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            ('groups.record_deleted', '=', 0),
+        ]
+
+    def __parse_qualified_column(self, column_name: str) -> AliasedColumn:
+        # TODO: This logic should be general and in the Query class.
+        # cannot do that yet since the column processing methods like
+        # column_expr do not have access to the Query yet.
+        match = self.QUALIFIED_COLUMN_REGEX.match(column_name)
+        if not match or not match[1] in [self.EVENTS_ALIAS, self.GROUPS_ALIAS]:
+            return (None, column_name)
```
Comment (tkaemming) [inline snuba/datasets/groups.py]: Not sure I understand the `not match[1] in [self.EVENTS_ALIAS, self.GROUPS_ALIAS]` part of this expression. Wouldn't this be an error condition if somebody provided an invalid table alias?

**Review by tkaemming — CHANGES_REQUESTED** [review]
Mostly questions or small changes, except for the `super` thing in the `Groups` constructor.

File: snuba/datasets/groups.py (hunk 11 above)
Comment (fpacifici) [reply snuba/datasets/groups.py]: The goal is to  still allow expressions like tags[], tags_key and so on to be used without alias to avoid making the events dataset know about the concept of alias. 
But, thinking over it 2 days later, I don't think it is a bad idea for the events dataset to know that a column name may have an alias in front. Will change it that way.

File: snuba/datasets/__init__.py [hunk 12]
```diff
@@ -59,12 +59,12 @@ def default_conditions(self):
         """
         return []
 
-    def column_expr(self, column_name, body):
+    def column_expr(self, column_name, body, table_alias: str=""):
```
Comment (fpacifici) [inline snuba/datasets/__init__.py]: I spent a very long time deciding if the default value should be "" or None. Finally in both cases the alias does not exists, avoiding a None input allows us to be more strict in type checking

File: snuba/datasets/groups.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+    def default_conditions(self):
+        return [
+            ('events.deleted', '=', 0),
+            ('groups.record_deleted', '=', 0),
+        ]
+
+    def column_expr(self, column_name, body, table_alias: str=""):
+        # Eventually joined dataset should not be represented by the same abstraction
+        # as joinable datasets. That will be easier through the TableStorage abstraciton.
```
Comment (tkaemming) [inline snuba/datasets/groups.py]:
```suggestion
        # as joinable datasets. That will be easier through the TableStorage abstraction.
```

File: snuba/util.py [hunk 14]
```diff
@@ -165,8 +187,11 @@ def column_expr(dataset, column_name, body, alias=None, aggregate=None):
     if aggregate:
         expr = function_expr(aggregate, expr)
 
-    alias = escape_col(alias or column_name)
-
+    use_escape_alias = state.get_config("use_escape_alias", 0)
+    # This is a precaution during the rollout.
```
Comment (fpacifici) [inline snuba/util.py]: This is the only non trivial change to the existing logic that can impact other datasets. Thus I will deploy this disabled, then once safe in prod we can quickly switch it on and off in case of issues. If everything is fine I will remove the check.

File: snuba/datasets/groups.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+            table_alias = match[1]
+            simple_column_name = match[2]
+            if table_alias == self.GROUPS_ALIAS:
+                return self.__grouped_message.column_expr(simple_column_name, body, table_alias)
+            elif table_alias == self.EVENTS_ALIAS:
+                return self.__events.column_expr(simple_column_name, body, table_alias)
+            else:
+                # This is probably an error condition. To keep consistency with the behavior
+                # in existing datasets, we let Clickhouse figure it out.
+                return super().column_expr(simple_column_name, body, table_alias)
```
Comment (tkaemming) [inline snuba/datasets/groups.py]: This function is much clearer now in general, thanks for commenting these details.

File: snuba/datasets/groups.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+            "Groups dataset cannot be referenced with table aliases. Alias provided {table_alias}"
+
+        match = self.QUALIFIED_COLUMN_REGEX.match(column_name)
+        if not match:
+            # anything that is not prefixed with a table alias is simply
+            # escaped and returned. It could be a literal.
+            return super().column_expr(column_name, body, table_alias)
+        else:
+            table_alias = match[1]
+            simple_column_name = match[2]
```
Comment (tkaemming) [inline snuba/datasets/groups.py]: Would it make sense to rewrite this as:

```
[table_alias, simple_column_name] = match
```

This should make it more obvious if the regular expression changes structurally in a way that could affect these assignments.

File: snuba/datasets/schemas/join.py [hunk 17]
```diff
@@ -70,6 +84,16 @@ class SubJoinSource(JoinedSource):
     def __str__(self) -> str:
         return f"{self.structure.get_data_source()}"
 
+    def get_schemas(self) -> Mapping[str, Schema]:
+        left = self.structure.left_source.get_schemas()
+        right = self.structure.right_source.get_schemas()
+        overlapping_aliases = left.keys() & right.keys()
```
Comment (tkaemming) [inline snuba/datasets/schemas/join.py]: I think this is one of my favorite Python 3 features outside of annotations.

File: snuba/datasets/schemas/join.py [hunk 18]
```diff
@@ -70,6 +84,16 @@ class SubJoinSource(JoinedSource):
...
+        for alias in overlapping_aliases:
+            # Ensures none defines the same alias twice in the join referring
+            # to different tables.
+            assert left[alias] == right[alias]
```
Comment (tkaemming) [inline snuba/datasets/schemas/join.py]: What's the case where you'd expect these to be equal?

File: snuba/query/query.py [hunk 19]
```diff
@@ -5,6 +5,7 @@
     Any,
     Mapping,
     MutableMapping,
+    NamedTuple,
```
Comment (tkaemming) [inline snuba/query/query.py]: 🤔 

I'm guessing this got left here by accident?

File: snuba/datasets/groups.py [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+            if table_alias == self.GROUPS_ALIAS:
+                return self.__grouped_message.column_expr(simple_column_name, body, table_alias)
```
Comment (fpacifici) [inline snuba/datasets/groups.py]: With this approach, composite datasets can parse the column name into table alias and column name. 
In theory the events dataset could do this parsing on its own but I opted for not doing it for two reasons:
1) risk: reduce the risk for this PR, so that this parsing logic only happens on a dataset not yet used in prod.
2) going ahead the parsing logic from the column_expr method is going to be separate from the formatting logic and closer to the validation. Finding table aliases will not be a dataset specific functionality but something generic so it would not end into events anyway.

With this approach the events dataset knows table aliases exist, but it does not to make assumptions on the name of the alias.

File: snuba/util.py [hunk 21]
```diff
@@ -55,21 +64,34 @@ def string_col(dataset, col):
...
+        col = ESCAPE_COL_RE.sub(r"\\\1", expr)
         return u'{}`{}`'.format(*NEGATE_RE.match(col).groups())
 
 
+def escape_alias(alias):
+    """
+    Alias escaping is different than column names when we introduce table aliases.
+    Using the column escaping function would consider "." safe, which is not for
+    an alias.
+    """
```
Comment (tkaemming) [inline snuba/util.py]: I'm torn on whether this comment should be closer to the regular expression definition or something.

File: snuba/util.py [hunk 22]
```diff
@@ -55,21 +64,34 @@ def string_col(dataset, col):
...
 
-def escape_col(col):
-    if not col:
-        return col
-    elif SAFE_COL_RE.match(col):
-        # Column is safe to use without wrapping.
-        return col
+def escape_expression(expr: Optional[str], regex) -> Optional[str]:
+    if not expr:
+        return expr
```
Comment (tkaemming) [inline snuba/util.py]: What's the case where this is `None`?

**Comment by fpacifici** [comment]
See the updated test plan in the description.

File: snuba/datasets/__init__.py (hunk 12 above)
Comment (tkaemming) [reply snuba/datasets/__init__.py]: > avoiding a None input allows us to be more strict in type checking

I don't understand the reasoning here. Can you elaborate?

Between the two, `None` makes more sense to me, but I think I'm likely missing something.

File: snuba/datasets/events.py [hunk 23]
```diff
@@ -32,6 +32,7 @@
     all_referenced_columns,
     escape_literal,
     string_col,
+    qualified_column,
```
Comment (tkaemming) [inline snuba/datasets/events.py]: The changes in this file are only reinforcing my thinking that datasets shouldn't be allowed to make these kind of query changes in the future — they should come from column definitions or extensions.

File: snuba/datasets/events.py [hunk 24]
```diff
@@ -355,23 +357,26 @@ def _tags_expr(self, column_name, body):
...
             val_list = u'arrayConcat([{}], {}.value)'.format(
                 ', '.join(string_col(self, p) for p in promoted),
-                col
+                qualified_col
             )
 
-        cols_used = all_referenced_columns(body) & set(['tags_key', 'tags_value'])
+        qualified_key = qualified_column("tags_key", table_alias)
+        qualified_value = qualified_column("tags_value", table_alias)
+        cols_used = all_referenced_columns(body) & set([qualified_key, qualified_value])
```
Comment (tkaemming) [inline snuba/datasets/events.py]: Is the assumption here that all of the columns in the body with `all_referenced_columns` are already qualified?

File: snuba/datasets/groups.py [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+            time_group_columns={
+                'events.time': 'events.timestamp',
+            },
+        )
+
+    def default_conditions(self):
+        return [
+            ('events.deleted', '=', 0),
+            ('groups.record_deleted', '=', 0),
+        ]
```
Comment (tkaemming) [inline snuba/datasets/groups.py]: This seems like it'll be a little brittle, hopefully this is temporary.

File: snuba/datasets/groups.py [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+            'project': ProjectExtension(),
+            'timeseries': TimeSeriesExtension(
+                default_granularity=3600,
+                default_window=timedelta(days=5),
+                timestamp_column='events.timestamp',
+            ),
+        }
+
+    def get_prewhere_keys(self) -> Sequence[str]:
+        return ['events.event_id', 'events.project_id']
```
Comment (tkaemming) [inline snuba/datasets/groups.py]: What's the thought process behind the prewhere keys for this dataset? (This isn't a leading question, I still don't really have a good handle on what good selection criteria are, I'm mostly curious.)

File: snuba/util.py (hunk 14 above)
Comment (tkaemming) [reply snuba/util.py]: Makes sense. Do you have anything specifically you'd expect to see if it goes bad (other than just "errors"?)

File: snuba/util.py [hunk 27]
```diff
@@ -45,6 +46,14 @@ def to_list(value):
     return value if isinstance(value, list) else [value]
 
 
+def qualified_column(column_name: str, alias: str="") -> str:
+    """
+    Returns a column in the form "table.column" if the table is not
+    empty. If the table is empty it returns the column itself.
+    """
+    return column_name if not alias else f"{alias}.{column_name}"
```
Comment (tkaemming) [inline snuba/util.py]: Do you think this will evolve into something where this is a data type in the `Query` class itself, rather than something that is applied functionally in various places? Or some other long-term plan?

**Review by tkaemming — APPROVED** [review]
A few questions inline, I don't think the answers would change the overall design in any way, mostly just curious. Nothing stands out to me that would be problematic, but the surface area here is obviously very large so I don't think there's much we can realistically do other than try to deploy it and see what happens.

File: snuba/datasets/events.py (hunk 23 above)
Comment (fpacifici) [reply snuba/datasets/events.py]: For "this kind of query changes" do you mean "everything that happens in column_expr" ?
If yes, then I agree, I wrote my theory about this topic in a previous comment (https://github.com/getsentry/snuba/pull/474#discussion_r326932220). I am planning to validate that approach in the next days.

File: snuba/datasets/events.py (hunk 24 above)
Comment (fpacifici) [reply snuba/datasets/events.py]: Yes. I would start with a strict syntax: in order to use join, everything has to be qualified. Then we can relax it if it makes sense. I will try to add an error condition to enforce this condition.

File: snuba/datasets/groups.py (hunk 25 above)
Comment (fpacifici) [reply snuba/datasets/groups.py]: It's from before I supported table_alias in the datasets, then I missed it. I can remove it right away.

File: snuba/datasets/groups.py (hunk 16 above)
Comment (fpacifici) [reply snuba/datasets/groups.py]: sure, will do.

File: snuba/datasets/groups.py (hunk 26 above)
Comment (fpacifici) [reply snuba/datasets/groups.py]: There was basically no thought process. I think I should change it at least to include all the events column. But finally I guess we will have to revisit how to optimize the query in a world with joins.

File: snuba/datasets/schemas/join.py (hunk 18 above)
Comment (fpacifici) [reply snuba/datasets/schemas/join.py]: (just an example, not a real join)

> events A LEFT JOIN groups B ON A.group_id = B.id LEFT JOIN group_assignee C ON A.group_id = C.group_id

Depending on how this is represented, the alias A referring to events will be present twice in the join clause

File: snuba/query/query.py (hunk 19 above)
Comment (fpacifici) [reply snuba/query/query.py]: yes

File: snuba/util.py (hunk 27 above)
Comment (fpacifici) [reply snuba/util.py]: yes definitively. 
The reason why we need this is that expressions in the current Snuba are passed around as string instead of being abstract data structures. Once we will move there, logic like this will be in the last stage of formatting the query.

File: snuba/util.py (hunk 22 above)
Comment (fpacifici) [reply snuba/util.py]: I am not really sure, I just picked it up from the previous escape_col to be safe. This is not used only in query processing but in other places as well.

File: snuba/util.py (hunk 14 above)
Comment (fpacifici) [reply snuba/util.py]: if something goes wrong here, most certainly the query will fail because it would be badly formatted.

**Comment by fpacifici** [comment]
_Nothing stands out to me that would be problematic, but the surface area here is obviously very large so I don't think there's much we can realistically do other than try to deploy it and see what happens._

Are you concerned about deploying this safely? Or about the design?
About the design, I expect some evolutions as soon as the product team starts playing with it (an example may be whether to enforce aliases everywhere), I am in contact with Chris for that.
About the production rollout, I added a kill switch where the change impacts existing datasets, I am open to protect the tag logic as well, if you feel it would be better, it does not seemed necessary to me because it is well covered with tests.

File: snuba/datasets/groups.py (hunk 16 above)
Comment (fpacifici) [reply snuba/datasets/groups.py]: Actually it does not work:
```
File "/Users/filippopacifici/snuba/snuba/datasets/groups.py", line 99, in column_expr
    [table_alias, simple_column_name] = match
TypeError: cannot unpack non-iterable re.Match object
```

Comment (tkaemming) [reply snuba/datasets/groups.py]: Oh, I guess this has to be `match.groups()`, but no big deal.

File: snuba/datasets/__init__.py (hunk 12 above)
Comment (tkaemming) [reply snuba/datasets/__init__.py]: To me, it seems like it'd be clearer for `None` to be the default and an empty string just seems like an invalid alias. I don't have a strong opinion about it either, and don't think it's necessarily worth changing unless it's likely to cause a problem with accidental usage (probably not, but I'm not as familiar with this as you are.)

File: snuba/datasets/events.py (hunk 23 above)
Comment (tkaemming) [reply snuba/datasets/events.py]: I think so, yes.

File: snuba/datasets/events.py (hunk 24 above)
Comment (tkaemming) [reply snuba/datasets/events.py]: Makes sense, thanks for the clarification. Probably worth asserting or at least commenting that here if you haven't already.

File: snuba/datasets/groups.py (hunk 26 above)
Comment (tkaemming) [reply snuba/datasets/groups.py]: Makes sense, thanks for clarification. It'd probably be helpful to comment this kind of thing in the future when it's being originally written to make sure it doesn't slip through by accident (not that's it's going to cause any significant problem, just that it's easily missed and might be misinterpreted as intentional later.)

File: snuba/datasets/schemas/join.py (hunk 18 above)
Comment (tkaemming) [reply snuba/datasets/schemas/join.py]: Ahh ok, this makes sense, thanks.

File: snuba/util.py (hunk 27 above)
Comment (tkaemming) [reply snuba/util.py]: This is also kind of what I was getting at with https://github.com/getsentry/snuba/pull/474/files/6a60ec84a117c6ecd640d7db3e07a17b00ed0507#r328877789.

File: snuba/util.py (hunk 22 above)
Comment (tkaemming) [reply snuba/util.py]: Hrm, ok. I guess that makes sense to be safe but it'd be helpful to note that the necessity is unclear for this kind of thing.

File: snuba/util.py (hunk 14 above)
Comment (tkaemming) [reply snuba/util.py]: Makes sense, thank you.

**Review by tkaemming — COMMENTED** [review]
Replied to a few of the questions, thanks for the clarifications.

File: snuba/datasets/groups.py (hunk 16 above)
Comment (fpacifici) [reply snuba/datasets/groups.py]: I did not try that, will check and fix it if it works.

File: snuba/util.py (hunk 22 above)
Comment (fpacifici) [reply snuba/util.py]: Ok, I will add a comment
